src/librustc_driver/profile/mod.rs

```diff
@@ -0,0 +1,167 @@
+                    },
+                    (ParseState::NoQuery,
+                     ProfileQueriesMsg::ProviderBegin) => {
+                        panic!("parse error: expected QueryBegin before beginning a provider")
+                    },
+                    (ParseState::NoQuery,
+                     ProfileQueriesMsg::ProviderEnd) => {
+                        let provider_extent = frame.traces;
+                        match stack.pop() {
+                            None =>
+                                panic!("parse error: expected a stack frame; found an empty stack"),
+                            Some(old_frame) => {
+                                match old_frame.parse_st {
+                                    ParseState::NoQuery =>
+                                        panic!("parse error: expected a stack frame for a query"),
+                                    ParseState::HaveQuery(q,start) => {
+                                        let duration = start.elapsed();
+                                        frame = StackFrame{
+                                            parse_st:ParseState::NoQuery,
+                                            traces:old_frame.traces
+                                        };
+                                        let trace = Rec {
+                                            effect: Effect::QueryBegin(q, CacheCase::Miss),
+                                            extent: Box::new(provider_extent),
+                                            start: start,
+                                            duration: duration,
+                                        };
+                                        frame.traces.push( trace );
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    // Parse State: HaveQuery
+                    (ParseState::HaveQuery(q,start),
+                     ProfileQueriesMsg::CacheHit) => {
+                        let duration = start.elapsed();
+                        let trace : Rec = Rec{
+                            effect: Effect::QueryBegin(q, CacheCase::Hit),
+                            extent: Box::new(vec![]),
+                            start: start,
+                            duration: duration,
+                        };
+                        frame.traces.push( trace );
+                        frame.parse_st = ParseState::NoQuery;
+                    },
+                    (ParseState::HaveQuery(_,_),
+                     ProfileQueriesMsg::ProviderBegin) => {
+                        stack.push(frame);
+                        frame = StackFrame{parse_st:ParseState::NoQuery, traces:vec![]};
+                    },
+                    (ParseState::HaveQuery(q,_),
+                     ProfileQueriesMsg::ProviderEnd) => {
+                        panic!("parse error: unexpected ProviderEnd;
```

Comment:
You can use a `\` to indicate that the string continues at the first non-whitespace character on the next line, like in:
```rust
    panic!("parse error: unexpected QueryBegin; \
            earlier query is unfinished: {:?} and now {:?}",
           q1, Query{span:span2, msg:querymsg2})
```